### ISSUE TITLE
Parse tree: Add placeholders to mark template substitutions

### ISSUE THREAD

**Issue by jpy-git** [sqlfluff/sqlfluff#2066]
When using Jinja templating the various templating constructs (e.g. if, for loops, etc) are stored in the parse tree as `[META] placeholder`
![image](https://user-images.githubusercontent.com/80432516/145119072-36a5cce9-73d9-49e5-b847-e529c66aa24a.png)

This is very useful when it comes to developing rules as it allows for for us to account for the existence of these in our logic and prevents unintentional rule flagging/potential corruptions.

The DBT templater on the other hand will often omit these constructs (namely `{{ config(materialized='view') }}` and `{% endmacro %}`) from the parse tree. This means that it's impossible for us to properly assess and fix lints such as L009 if the file contains one of these constructs.

e.g. in #2046 a file ends with
```
ORDER BY rl.RunID

{% endmacro %}
```
and parses to
```
[L:  8, P:  1]      |                        identifier:                           'RunID'
[L:  8, P: 44]      |                    [META] dedent:
[L:  8, P:  1]      |    newline:                                                  '\n'
[L:  8, P:  1]      |    newline:                                                  '\n'
[L:  8, P: 44]      |    newline:                                                  '\n'
```
so from the perspective of L009 the file ends with 3 trailing newlines and hence raises the lint error. However, this query should really parse to something like:
```
[L:  8, P:  1]      |                        identifier:                           'RunID'
[L:  8, P: 44]      |                    [META] dedent:
[L:  8, P:  1]      |    newline:                                                  '\n'
[L:  8, P:  1]      |    newline:                                                  '\n'
[L:  8, P:  1]      |    [META] dedent:
[L:  8, P:  1]      |    [META] placeholder:                             [Type: 'macro_end', Raw: '{% endmacro %}']
[L:  8, P: 44]      |    newline:                                                  '\n'
```
This would allow us to properly account for these templating constructs in the rules, like we are able to do when using Jinja.

When encountering this same issue for L052 ([here](https://github.com/sqlfluff/sqlfluff/pull/1840#issuecomment-962682902)) we were able to implement some basic logic to at least abandon the lint/fix if one of these potential invisible dbt constructs were detected, but it's not a full fix as we would have to start implementing this sort of thing in all rules and prevents us from fixing a lot of situations for dbt that would be simple if given the correct parse tree.

## Dialect
ANSI

## Version
SQLFluff 0.8.2 (HEAD of main branch)
Python 3.9.9
Impacts all dbt versions.

## Configuration
Default config

**Comment by jpy-git** [comment]
@barrywhart I was looking through old PRs to see if we've had other occurences of the above issue and came across your PR #1907. The issue you describe and fix within it sounds very similar to these other DBT tags that are being filtered out in the pre-processing step. Would you be able to take a look at this and check whether that fix can be extended to these other tags please? 😄

**Comment by barrywhart** [comment]
@jpy-git: I'll add it to my list. I've been pretty distracted with job interviews the past month or so, but that's starting to settle down, and I expect I can work on the project next week. (In fact, I was just working today to set up a better development environment -- my new notebook is an M1 Mac, which has some compatibility issues with various dbt dependencies.) 

My (very vague, possibly wrong) impression is that placeholders are created for template tags `{{ }}` but only if they generate some code (i.e. if they render as the empty string, no placeholder), also that template control structures `{% %}` do not generate placeholders. Does this sound correct to you, @alanmcruickshank?

We might encounter a few complications with changing this, but it could be a worthwhile investment -- dbt is taking over the world, and it seems like SQLFluff should be able to handle the amount and complexity of templating typical of dbt projects.

**Comment by jpy-git** [comment]
@barrywhart awesome thanks, it's not urgent and likely best to wait until dbt 1.0.0 is working first 😄 But, as you said, with dbt becoming more popular it would be good to get a proper fix for this so just pulling together the relevant info 👍

**Comment by jpy-git** [comment]
> * There's an existing config setting called `ignore_templated_areas` (mentioned [here](https://docs.sqlfluff.com/en/stable/configuration.html#macro-templating), but not well documented) which should _maybe_ handle this already. Is this a bug? See the function `Linter.remove_templated_errors()` for the implementation of `ignore_templated_areas`.

So the issues I had in [L050](https://github.com/sqlfluff/sqlfluff/pull/1840#issuecomment-962682902) with that `remove_templated_errors` method is because we would have a parse tree with multiple leading newlines but no reference to a `{{ config(materialized='view') }}` somewhere in there. So the rule would suggest deleting those newline segments.
However, since `{{ config(materialized='view') }}` evaluated to blank and gave no meta segment (it disappeared between the raw and the source slicing) those newline segments would not be contiguous when referring to their raw position (i.e. there's a gap in the raw positions where that config substitution was). So if you were to check `is_literal` on all those newlines they would all be true as they aren't from the config substitution and hence the fix is not aborted.
Basically `remove_templated_errors` can only work if the segment it's checking came from a template substitution, not if there is some invisible template between the segments. 

[This](https://github.com/sqlfluff/sqlfluff/blob/ad12313b91f461c53a3e8759ea9dce559c918f26/src/sqlfluff/rules/L050.py#L65) is the logic I had to use to infer these invisible segments in L050 so we can incorporate some of this into `ignore_templated_areas` to catch this type of edge cases. That should help stop these fixes like in #2046 from being flagged in general.

**Comment by jpy-git** [comment]
My main concern with approach 1 is we're potentially throwing away a lot of changes where we could fix
e.g. L050: we could have 100 newlines and 1 `{{ config(materialized='view') }}` halfway down and we would have to throw away that fix, whereas if we had a placeholder we can just keep the rule logic as `is not whitespace` to find the first non-whitespace (including the config sub).

If we were to consider option 2, I'd envision in the parse tree like:
```
[META] placeholder:                             [Type: 'substitution_start', Raw: '{{ <raw_content_here> }}']
<the substituted code segments here>
[META] placeholder:                             [Type: 'substitution_end']
```

**Comment by barrywhart** [comment]
Makes sense. If all goes well, I can start looking at this in the next week or so.

I think the parser change will be easy, the main question is whether it will raise any gotchas. 🤞🏽

**Comment by jpy-git** [comment]
@barrywhart should we close this issue given that we aren't taking this approach going forwards?

**Comment by barrywhart** [comment]
@jpy-git: I'm happy to close it.  I wonder, though, do you want to create a similar issue expressed in terms of the new API-based work on issue #1808? Because it definitely seems like something we want to handle or improve on, only the *design* of how to do it has changed.

**Comment by jpy-git** [comment]
Yeah let me create a new issue rephrasing the template detection issue in terms of the API-based approach and will close this one after 👍

**Comment by barrywhart** [comment]
@jpy-git: Have you created the related issue yet? Wondering if we can close this one.
